### PR TITLE
storage/drivers/ceph: Differentiate between a true non-existent OSD pool and a Ceph internal issue

### LINKED ISSUE #12163: Ceph storage create pool has a possibility of deleting exist pool
# Required information

 * Distribution: Ubuntu
 * Distribution version: 22.04
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.15.0-78-generic
   * LXC version: 5.15
   * LXD version: 5.15
   * Storage backend in use: ceph

# Issue description
If someone try to create a ceph storage which has the same name as an exist one, the exist pool may be deleted

here checks whether ceph pool is exist
[check pool](https://github.com/canonical/lxd/blob/c77ee90f5032f862e4a514bed76ee736693ba64d/lxd/storage/drivers/driver_ceph.go#L153)

the osdPoolExists() checks pool exist by cmd return
[osdPoolExists 1](https://github.com/canonical/lxd/blob/c77ee90f5032f862e4a514bed76ee736693ba64d/lxd/storage/drivers/driver_ceph_utils.go#L49)

This means if there’s a temp error (network jitter or any other reason) about ceph while osdPoolExists() querying, lxd will try to create pool.
And if the error fix at a short time, lxd will create fail beacuse of the name of pool is exist, and delete the exist pool by revert?

# Steps to reproduce

 1. create a pool named ‘remote3’: lxc storage create remote3 ceph --verbose
 2. change my ceph key to a wrong value
 3. create a pool named ‘remote3’ again: lxc storage create remote3 ceph --verbose
 4. set correct ceph key immediately, and ‘remote3’ will be delete by lxd.
Deleting will also happen even if I set ctrl+C to stop lxd commend after step 3. If I do this, there’s even no error log on my terminal

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes #12163

### CODE REVIEW

File: lxd/storage/drivers/driver_ceph_utils.go [hunk 1]
```diff
@@ -57,7 +57,24 @@ func (d *ceph) osdPoolExists() bool {
 		d.config["ceph.osd.pool_name"],
 		"size")
 
-	return err == nil
+	if err != nil {
+		runErr, ok := err.(shared.RunError)
```
Comment (tomponline) [inline lxd/storage/drivers/driver_ceph_utils.go:61]: Would it be easier to use shared.ExitStatus() here?

File: lxd/storage/drivers/driver_ceph.go [hunk 2]
```diff
@@ -246,7 +250,11 @@ func (d *ceph) Create() error {
 // Delete removes the storage pool from the storage device.
 func (d *ceph) Delete(op *operations.Operation) error {
 	// Test if the pool exists.
-	poolExists := d.osdPoolExists()
+	poolExists, err := d.osdPoolExists()
+	if err != nil {
+		return err
```
Comment (tomponline) [inline lxd/storage/drivers/driver_ceph.go:255]: And we should wrap error with context here too.

More generally we should wrap context when the error won't make much sense to the user without it.

File: test/suites/storage_driver_ceph.sh [hunk 3]
```diff
@@ -40,6 +40,22 @@ test_storage_driver_ceph() {
     lxc storage create "lxdtest-$(basename "${LXD_DIR}")-valid-ceph-pool-config" ceph volume.block.filesystem=ext4 volume.block.mount_options=discard volume.size=25MiB ceph.rbd.clone_copy=true ceph.osd.pg_num=16
     lxc storage delete "lxdtest-$(basename "${LXD_DIR}")-valid-ceph-pool-config"
 
+    # Test that if there is a temporay error on the Ceph side (network jitter, wrong auth key, etc.)
+    # while creating a storage pool that already exists, lxd will not delete this storage pool.
+    # Step 1: Create a new Ceph storage pool
+    lxc storage create "lxdtest-$(basename "${LXD_DIR}")-pool3" ceph volume.size=25MiB ceph.osd.pg_num=16
+    # Step 2: Change the Ceph admin client key to a wrong value in /var/snap/microceph/current/conf/ceph.keyring
+    original_key=$(awk -F "key = " '/key/ {print $2}' /var/snap/microceph/current/conf/ceph.keyring)
```
Comment (tomponline) [inline test/suites/storage_driver_ceph.sh:48]: Sorry, we cannot have the test suite depend on microceph, as the test suite is used by non-ubuntu and non-microceph users.

Comment (tomponline) [reply test/suites/storage_driver_ceph.sh:48]: I'm not sure we can test this well in the test suite as we do not control the ceph environment.

Comment (gabrielmougard) [reply test/suites/storage_driver_ceph.sh:48]: Yes, I was wondering how to do this cleanly but I clearly ran out of ideas...

Comment (tomponline) [reply test/suites/storage_driver_ceph.sh:48]: lets take the test parts that depend on microceph out for now.

Comment (gabrielmougard) [reply test/suites/storage_driver_ceph.sh:48]: If we are taking them out, then I don't see the point of this test at all as the main purpose is to simulate some kind of Ceph outage.. shall we just remove this test commit ?

Comment (tomponline) [reply test/suites/storage_driver_ceph.sh:48]: yep ok with me

**Pull request merged**